File: sdk/typescript/src/rpc/http.ts
/**
 * HTTP JSON-RPC 2.0 transport (fetch-based) with retries & AbortSignal support.
 *
 * Usage:
 *   import { createHttpClient } from '@animica/sdk/rpc'
 *   const rpc = createHttpClient('http://localhost:8545/rpc')
 *   const head = await rpc.request<Head>('chain.getHead')
 *
 * Features:
 *  - Exponential backoff with jitter (429/5xx aware, honors Retry-After)
 *  - Per-attempt timeout and external AbortSignal
 *  - Batch requests with stable ordering
 *  - Strong typing for request/response envelopes
 */

import {
  RpcTransport,
  RequestOptions,
  JsonRpcResponse,
  isJsonRpcFailure,
  isJsonRpcSuccess
} from './index'
import {
  fetchWithRetry,
  HttpError,
  AbortError,
  mergeSignals,
  type FetchRetryOptions
} from '../utils/retry'
import { RpcError } from '../errors'

/** Options for the HTTP JSON-RPC client. */
export interface HttpClientOptions {
  /** Extra headers to send with every request (e.g., API keys). */
  headers?: Record<string, string>
  /**
   * Retry settings. Enable/disable and tune backoff/jitter.
   * Defaults: enabled=true, minDelay=200ms, factor=2, maxDelay=10s, full jitter, retries=5.
   */
  retry?: (FetchRetryOptions & { enabled?: boolean }) | undefined
  /**
   * Default per-attempt timeout (ms). Can be overridden per call via RequestOptions.timeoutMs.
   * Default: 15_000 ms.
   */
  timeoutMs?: number
  /**
   * Function to create JSON-RPC ids. Default is an incrementing integer counter.
   */
  idFactory?: () => number | string | null
}

/** Callable tuple used by batch(). */
export interface RpcCall<P = unknown> {
  method: string
  params?: P
}

/** Factory to create an HTTP JSON-RPC transport bound to a base URL. */
export function createHttpClient(baseUrl: string, opts?: HttpClientOptions): HttpClient {
  return new HttpClient(baseUrl, opts)
}

export class HttpClient implements RpcTransport {
  private baseUrl: string
  private headers: Record<string, string>
  private retry: FetchRetryOptions & { enabled: boolean }
  private timeoutMs: number
  private idFactory: () => number | string | null
  private seq = 1

  constructor(baseUrl: string, opts?: HttpClientOptions) {
    this.baseUrl = normalizeUrl(baseUrl)
    this.headers = {
      'content-type': 'application/json',
      accept: 'application/json',
      ...(opts?.headers ?? {})
    }
    const r = opts?.retry ?? {}
    this.retry = {
      enabled: r.enabled !== false,
      retries: r.retries ?? 5,
      minDelay: r.minDelay ?? 200,
      factor: r.factor ?? 2,
      maxDelay: r.maxDelay ?? 10_000,
      jitter: r.jitter ?? 'full',
      honorRetryAfter: r.honorRetryAfter ?? true,
      retryMethods: r.retryMethods ?? ['POST'] // JSON-RPC POSTs are idempotent per id; server must de-duplicate
    }
    this.timeoutMs = opts?.timeoutMs ?? 15_000
    this.idFactory = opts?.idFactory ?? (() => this.seq++)
  }

  /** Perform a single JSON-RPC request and return the typed result. */
  async request<R = unknown, P = unknown>(method: string, params?: P, opts?: RequestOptions): Promise<R> {
    const id = this.idFactory()
    const payload = makeSinglePayload(method, params, id)
    const signal = mergeSignals([opts?.signal])

    const response = await fetchWithRetry(
      this.baseUrl,
      {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(payload),
        signal
      },
      this.retry.enabled
        ? { ...this.retry, attemptTimeoutMs: opts?.timeoutMs ?? this.timeoutMs, signal }
        : { retries: 0, attemptTimeoutMs: opts?.timeoutMs ?? this.timeoutMs, signal }
    )

    const json = await parseJson(response)

    // Some servers may respond with a batch even for single requests; handle permissively.
    if (Array.isArray(json)) {
      const match = json.find((x) => (x as any)?.id === id) as JsonRpcResponse<R> | undefined
      if (!match) throw new RpcError(-32603, 'Invalid batch response: missing id for single request')
      if (isJsonRpcSuccess<R>(match)) return match.result
      if (isJsonRpcFailure(match)) throw new RpcError(match.error.code, match.error.message, match.error.data)
      throw new RpcError(-32603, 'Invalid JSON-RPC response shape')
    }

    const obj = json as JsonRpcResponse<R>
    if (isJsonRpcSuccess<R>(obj)) return obj.result
    if (isJsonRpcFailure(obj)) throw new RpcError(obj.error.code, obj.error.message, obj.error.data)
    throw new RpcError(-32603, 'Invalid JSON-RPC response shape')
  }

  /**
   * Perform a batch of JSON-RPC calls.
   * Returns results in the same order as calls[].
   */
  async batch(calls: ReadonlyArray<RpcCall>, opts?: RequestOptions): Promise<unknown[]> {
    if (!Array.isArray(calls) || calls.length === 0) return []
    const ids = calls.map(() => this.idFactory())
    const payload = calls.map((c, i) => makeSinglePayload(c.method, c.params, ids[i]))

    const signal = mergeSignals([opts?.signal])
    const response = await fetchWithRetry(
      this.baseUrl,
      {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(payload),
        signal
      },
      this.retry.enabled
        ? { ...this.retry, attemptTimeoutMs: opts?.timeoutMs ?? this.timeoutMs, signal }
        : { retries: 0, attemptTimeoutMs: opts?.timeoutMs ?? this.timeoutMs, signal }
    )

    const json = await parseJson(response)
    if (!Array.isArray(json)) throw new RpcError(-32603, 'Batch request expected array response')

    // Map responses by id for stable ordering
    const byId = new Map<any, JsonRpcResponse>()
    for (const r of json) byId.set((r as any)?.id, r)

    return ids.map((id) => {
      const item = byId.get(id) as JsonRpcResponse | undefined
      if (!item) throw new RpcError(-32603, `Missing response for id ${String(id)}`)
      if (isJsonRpcSuccess(item)) return item.result
      if (isJsonRpcFailure(item)) throw new RpcError(item.error.code, item.error.message, item.error.data)
      throw new RpcError(-32603, 'Invalid JSON-RPC response shape in batch')
    })
  }

  /** Create a shallow-cloned client with additional headers (e.g., new API key). */
  withHeaders(headers: Record<string, string>): HttpClient {
    return new HttpClient(this.baseUrl, {
      headers: { ...this.headers, ...headers },
      retry: this.retry,
      timeoutMs: this.timeoutMs,
      idFactory: this.idFactory
    })
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Helpers
// ──────────────────────────────────────────────────────────────────────────────

function makeSinglePayload<P = unknown>(method: string, params: P | undefined, id: number | string | null) {
  const payload: any = {
    jsonrpc: '2.0',
    method,
    id
  }
  if (params !== undefined) payload['params'] = params
  return payload
}

async function parseJson(res: Response): Promise<unknown> {
  const text = await res.text()
  try {
    // JSON-RPC forbids BigInt, so plain JSON.parse is fine here.
    return JSON.parse(text)
  } catch (e) {
    // Surface richer diagnostics while preserving original HTTP context.
    const hint = text && text.length < 2048 ? `; body="${text}"` : ''
    throw new HttpError(`Invalid JSON from server${hint}`, res.status, res.statusText, text)
  }
}

function normalizeUrl(u: string): string {
  // Accept base URLs with or without trailing slashes; do not mutate query/hash.
  // For JSON-RPC we almost always POST to the exact baseUrl.
  return u.replace(/\s+/g, '')
}

export type { FetchRetryOptions } from '../utils/retry'
export { HttpError, AbortError } from '../utils/retry'
export { RpcError } from '../errors'
export default createHttpClient

File: sdk/rust/src/rpc/http.rs
//! Minimal, robust JSON-RPC HTTP client with retries (reqwest).
//!
//! Features:
//! - Async `reqwest` client with sane defaults (timeouts, UA).
//! - Exponential backoff with jitter for transient failures (5xx/429/timeouts).
//! - Typed single-call API and convenient raw/batch helpers.
//! - Optional bearer auth & custom headers.
//!
//! This client is transport-only. It does not interpret chain semantics.

use crate::error::{Error, Result};
use reqwest::{header, Client, StatusCode, Url};
use serde::{de::DeserializeOwned, Serialize};
use serde_json::{json, Value};
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Duration;

/// JSON-RPC 2.0 request envelope.
#[derive(Debug, Serialize)]
struct RpcRequest<'a> {
    jsonrpc: &'static str,
    id: u64,
    method: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    params: Option<Value>,
}

/// JSON-RPC 2.0 response envelope.
#[derive(Debug, Deserialize)]
struct RpcResponse<T> {
    jsonrpc: String,
    id: Value,
    #[serde(default)]
    result: Option<T>,
    #[serde(default)]
    error: Option<RpcErrorObj>,
}

#[derive(Debug, Deserialize)]
struct RpcErrorObj {
    code: i64,
    message: String,
    #[serde(default)]
    data: Option<Value>,
}

/// Builder for [`HttpClient`].
#[derive(Clone, Debug)]
pub struct HttpClientBuilder {
    endpoint: Url,
    timeout: Duration,
    connect_timeout: Duration,
    max_retries: u32,
    retry_base: Duration,
    default_headers: header::HeaderMap,
    user_agent: Option<String>,
}

impl HttpClientBuilder {
    pub fn new(endpoint: Url) -> Self {
        Self {
            endpoint,
            timeout: Duration::from_secs(20),
            connect_timeout: Duration::from_secs(10),
            max_retries: 3,
            retry_base: Duration::from_millis(250),
            default_headers: header::HeaderMap::new(),
            user_agent: None,
        }
    }

    pub fn from_str(endpoint: &str) -> Result<Self> {
        Ok(Self::new(endpoint.parse::<Url>().map_err(|e| Error::Transport(format!("bad URL: {e}")))?))
    }

    pub fn timeout(mut self, timeout: Duration) -> Self {
        self.timeout = timeout;
        self
    }

    pub fn connect_timeout(mut self, timeout: Duration) -> Self {
        self.connect_timeout = timeout;
        self
    }

    pub fn max_retries(mut self, retries: u32) -> Self {
        self.max_retries = retries;
        self
    }

    pub fn retry_base(mut self, base: Duration) -> Self {
        self.retry_base = base;
        self
    }

    pub fn bearer_auth(mut self, token: &str) -> Self {
        let value = header::HeaderValue::from_str(&format!("Bearer {token}"))
            .unwrap_or_else(|_| header::HeaderValue::from_static("Bearer INVALID"));
        self.default_headers.insert(header::AUTHORIZATION, value);
        self
    }

    pub fn header(mut self, key: header::HeaderName, value: header::HeaderValue) -> Self {
        self.default_headers.insert(key, value);
        self
    }

    pub fn user_agent(mut self, ua: &str) -> Self {
        self.user_agent = Some(ua.to_owned());
        self
    }

    pub fn build(self) -> Result<HttpClient> {
        let mut headers = self.default_headers.clone();
        headers.entry(header::CONTENT_TYPE).or_insert(header::HeaderValue::from_static("application/json"));
        if let Some(ua) = self.user_agent {
            headers
                .entry(header::USER_AGENT)
                .or_insert(header::HeaderValue::from_str(&ua).unwrap_or_else(|_| header::HeaderValue::from_static("animica-rust-sdk")));
        } else {
            headers
                .entry(header::USER_AGENT)
                .or_insert(header::HeaderValue::from_static("animica-rust-sdk"));
        }

        let client = Client::builder()
            .default_headers(headers)
            .connect_timeout(self.connect_timeout)
            .timeout(self.timeout)
            .pool_idle_timeout(Duration::from_secs(30))
            .tcp_nodelay(true)
            .build()
            .map_err(|e| Error::Transport(format!("reqwest build: {e}")))?;

        Ok(HttpClient {
            endpoint: self.endpoint,
            client,
            max_retries: self.max_retries,
            retry_base: self.retry_base,
            id: AtomicU64::new(1),
        })
    }
}

/// Async JSON-RPC HTTP client.
#[derive(Clone)]
pub struct HttpClient {
    endpoint: Url,
    client: Client,
    max_retries: u32,
    retry_base: Duration,
    id: AtomicU64,
}

impl std::fmt::Debug for HttpClient {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("HttpClient")
            .field("endpoint", &self.endpoint)
            .field("max_retries", &self.max_retries)
            .finish()
    }
}

impl HttpClient {
    /// Quick constructor with defaults.
    pub fn new(endpoint: &str) -> Result<Self> {
        HttpClientBuilder::from_str(endpoint)?.build()
    }

    /// Create a builder for custom configuration.
    pub fn builder(endpoint: &str) -> Result<HttpClientBuilder> {
        HttpClientBuilder::from_str(endpoint)
    }

    /// Perform a typed JSON-RPC call with automatic retries on transient errors.
    pub async fn call<T, P>(&self, method: &str, params: P) -> Result<T>
    where
        T: DeserializeOwned,
        P: Serialize,
    {
        let params_value = Some(serde_json::to_value(params).map_err(|e| Error::Serde(format!("params: {e}")))?);
        self.call_value::<T>(method, params_value).await
    }

    /// Same as [`call`] but takes pre-built `serde_json::Value` for params (or `None`).
    pub async fn call_value<T>(&self, method: &str, params: Option<Value>) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let id = self.next_id();
        let req = RpcRequest {
            jsonrpc: "2.0",
            id,
            method,
            params,
        };
        let body = serde_json::to_vec(&req).map_err(|e| Error::Serde(format!("encode request: {e}")))?;
        // Retry loop
        let mut last_err: Option<Error> = None;
        for attempt in 0..=self.max_retries {
            match self.try_send::<T>(&body).await {
                Ok(v) => return Ok(v),
                Err(e) => {
                    if !self.should_retry(&e) || attempt == self.max_retries {
                        return Err(e);
                    }
                    last_err = Some(e);
                    self.sleep_backoff(attempt).await;
                }
            }
        }
        Err(last_err.unwrap_or_else(|| Error::Transport("unreachable retry loop".into())))
    }

    /// Perform a **raw** call returning the untyped `serde_json::Value` result.
    pub async fn call_raw(&self, method: &str, params: Option<Value>) -> Result<Value> {
        self.call_value::<Value>(method, params).await
    }

    /// Execute a JSON-RPC batch. Each item is `(method, params)`.
    /// Returns a vector of results **ordered by request id**, not input order (JSON-RPC spec).
    pub async fn batch(&self, calls: Vec<(&str, Option<Value>)>) -> Result<Vec<Result<Value>>> {
        if calls.is_empty() {
            return Ok(vec![]);
        }
        // Build batch with strictly increasing ids.
        let mut next = self.next_id();
        let reqs: Vec<RpcRequest<'_>> = calls
            .iter()
            .map(|(m, p)| {
                let r = RpcRequest {
                    jsonrpc: "2.0",
                    id: next,
                    method: *m,
                    params: p.clone(),
                };
                next = next.saturating_add(1);
                r
            })
            .collect();

        let body = serde_json::to_vec(&reqs).map_err(|e| Error::Serde(format!("encode batch: {e}")))?;

        // Retry loop
        let mut last_err: Option<Error> = None;
        for attempt in 0..=self.max_retries {
            match self.try_send_batch(&body).await {
                Ok(v) => return Ok(v),
                Err(e) => {
                    if !self.should_retry(&e) || attempt == self.max_retries {
                        return Err(e);
                    }
                    last_err = Some(e);
                    self.sleep_backoff(attempt).await;
                }
            }
        }
        Err(last_err.unwrap_or_else(|| Error::Transport("unreachable retry loop".into())))
    }

    // --------------------------- internals ----------------------------------

    fn next_id(&self) -> u64 {
        self.id.fetch_add(1, Ordering::Relaxed)
    }

    async fn try_send<T>(&self, body: &[u8]) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let resp = self
            .client
            .post(self.endpoint.clone())
            .body(body.to_vec())
            .send()
            .await
            .map_err(|e| Error::Transport(format!("send: {e}")))?;

        let status = resp.status();
        let bytes = resp.bytes().await.map_err(|e| Error::Transport(format!("read body: {e}")))?;
        if !status.is_success() {
            return Err(http_status_error(status, &bytes));
        }

        let parsed: RpcResponse<T> = serde_json::from_slice(&bytes)
            .map_err(|e| Error::Serde(format!("decode rpc response: {e}; body={}", truncate_body(&bytes))))?;

        if let Some(err) = parsed.error {
            return Err(Error::Rpc(err.code, if let Some(data) = err.data {
                format!("{} | data={}", err.message, data)
            } else {
                err.message
            }));
        }
        parsed
            .result
            .ok_or_else(|| Error::Rpc(-32603, "missing result and error".into()))
    }

    async fn try_send_batch(&self, body: &[u8]) -> Result<Vec<Result<Value>>> {
        let resp = self
            .client
            .post(self.endpoint.clone())
            .body(body.to_vec())
            .send()
            .await
            .map_err(|e| Error::Transport(format!("send batch: {e}")))?;

        let status = resp.status();
        let bytes = resp.bytes().await.map_err(|e| Error::Transport(format!("read body: {e}")))?;
        if !status.is_success() {
            return Err(http_status_error(status, &bytes));
        }

        let parsed: Vec<RpcResponse<Value>> = serde_json::from_slice(&bytes)
            .map_err(|e| Error::Serde(format!("decode batch response: {e}; body={}", truncate_body(&bytes))))?;

        // Map to ordered results by id ascending (as many servers already do).
        let mut items: Vec<(u64, Result<Value>)> = Vec::with_capacity(parsed.len());
        for r in parsed {
            // best-effort id → u64
            let id_num = r
                .id
                .as_u64()
                .or_else(|| r.id.as_str().and_then(|s| s.parse::<u64>().ok()))
                .unwrap_or(0);

            let res = if let Some(err) = r.error {
                Err(Error::Rpc(
                    err.code,
                    if let Some(data) = err.data {
                        format!("{} | data={}", err.message, data)
                    } else {
                        err.message
                    },
                ))
            } else if let Some(val) = r.result {
                Ok(val)
            } else {
                Err(Error::Rpc(-32603, "missing result and error".into()))
            };
            items.push((id_num, res));
        }
        items.sort_by_key(|(id, _)| *id);
        Ok(items.into_iter().map(|(_, r)| r).collect())
    }

    fn should_retry(&self, err: &Error) -> bool {
        match err {
            Error::Transport(msg) => {
                // Rough heuristic: timeouts, DNS/connect, temporary network issues
                contains_any(msg, &["timeout", "timed out", "connect", "tls", "closed", "reset", "EOF"])
            }
            Error::Http(status, _msg) => {
                // Retry on 408, 425, 429, 5xx
                *status == StatusCode::REQUEST_TIMEOUT
                    || *status == StatusCode::TOO_EARLY
                    || *status == StatusCode::TOO_MANY_REQUESTS
                    || status.is_server_error()
            }
            Error::Rpc(code, _msg) => {
                // Retry internal server errors / rate-limit style custom codes.
                *code == -32000 || *code == -32001 || *code == -32002
            }
            _ => false,
        }
    }

    async fn sleep_backoff(&self, attempt: u32) {
        // attempt = 0 → base, 1 → 2x, etc., capped to 3s
        let base = self.retry_base.as_millis() as u64;
        let pow = 1u64.saturating_shl(attempt.min(6)); // cap growth
        let max_ms = (base.saturating_mul(pow)).min(3_000);
        let jitter = fastrand::u64(0..=max_ms / 2);
        let dur = Duration::from_millis(max_ms / 2 + jitter);
        tokio::time::sleep(dur).await
    }
}

// --------------------------- helpers -----------------------------------------

fn truncate_body(bytes: &[u8]) -> String {
    const LIM: usize = 512;
    let s = String::from_utf8_lossy(bytes);
    if s.len() > LIM {
        format!("{}...[+{}B]", &s[..LIM], s.len() - LIM)
    } else {
        s.into_owned()
    }
}

fn http_status_error(status: StatusCode, body: &[u8]) -> Error {
    let snippet = truncate_body(body);
    Error::Http(status, format!("http {}: {}", status.as_u16(), snippet))
}

fn contains_any(haystack: &str, needles: &[&str]) -> bool {
    let hs = haystack.to_ascii_lowercase();
    needles.iter().any(|n| hs.contains(&n.to_ascii_lowercase()))
}

// ------------------------------ tests ----------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builder_defaults() {
        let b = HttpClient::builder("http://localhost:8545").unwrap();
        let c = b.build().unwrap();
        assert_eq!(c.max_retries, 3);
    }

    #[test]
    fn id_increments() {
        let c = HttpClient::builder("http://localhost:8545").unwrap().build().unwrap();
        let a = c.next_id();
        let b = c.next_id();
        assert!(b > a);
    }

    #[test]
    fn retry_heuristics() {
        let c = HttpClient::builder("http://localhost:8545").unwrap().build().unwrap();
        assert!(c.should_retry(&Error::Transport("connection reset by peer".into())));
        assert!(c.should_retry(&Error::Http(StatusCode::INTERNAL_SERVER_ERROR, "oops".into())));
        assert!(!c.should_retry(&Error::Rpc(-32601, "method not found".into())));
    }
}
